README.md
# Alipay

Unofficial Alipay ruby gem.

Note: Alipay::Client API does not have enough feedback in production yet, please fully test in your staging environment before production. You can find legacy API document [here](doc/legacy_api.md).

You should read [https://doc.open.alipay.com](https://doc.open.alipay.com) before using this gem.

## Installation

To install using [Bundler](http://bundler.io/). Add this line to your
application's Gemfile:

```ruby
gem 'alipay'
```

Then run:
```bash
$ bundle
```

Or you can manually install using [RubyGems](http://rubygems.org/):
```bash
$ gem install alipay
```

## Getting Started

This gem needs to be configured with your application's private key for Alipay and Alipay's public key. Here is a [quick guide](doc/rsa_key_en.md) on generating RSA key for use with this gem to get you started.

### Setup
```ruby
require 'alipay'

# setup the client to communicate with either production API or sandbox API
# https://openapi.alipay.com/gateway.do (Production)
# https://openapi.alipaydev.com/gateway.do (Sandbox)
API_URL =  'https://openapi.alipaydev.com/gateway.do'

# setup your own credentials and certificates
APP_ID = '2016xxxxxxxxxxxx'
APP_PRIVATE_KEY="-----BEGIN RSA PRIVATE KEY-----\nxkbt...4Wt7tl\n-----END RSA PRIVATE KEY-----\n"
ALIPAY_PUBLIC_KEY="-----BEGIN PUBLIC KEY-----\nTq43T5...OVUAQb3R\n-----END PUBLIC KEY-----\n"

# initialize a client to communicate with the Alipay API
@alipay_client = Alipay::Client.new(
  url: API_URL,
  app_id: APP_ID,
  app_private_key: APP_PRIVATE_KEY,
  alipay_public_key: ALIPAY_PUBLIC_KEY
)
```

### Create a payment
```ruby
@alipay_client.page_execute_url(
  method: 'alipay.trade.page.pay',
  biz_content: JSON.generate({
    out_trade_no: '20160401000000',
    product_code: 'FAST_INSTANT_TRADE_PAY',
    total_amount: '0.01',
    subject: 'test'
  }, ascii_only: true), # ascii_only is important!
  timestamp: '2016-04-01 00:00:00'
)

# This method will then return a payment url
# => 'https://openapi.alipaydev.com/gateway.do?app_id=201600...'
```

Read [Alipay::Client](lib/alipay/client.rb) or the [Quick Start Guide](doc/quick_start_en.md) for usage detail.

## Contributing

Bug report or pull request are welcome.

### Make a pull request

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

Please write unit test with your code if necessary.

## License

MIT License


CHANGELOG.md
## master

Newer release changelog, see https://github.com/chloerei/alipay/releases

## v0.16.0 (2020-05-15)

- Support certificate signature, thanks @moohao #92

## v0.15.2 (2019-08-02)

- page_execute_form invalid-signature charset, thanks @lingceng #90

## v0.15.1 (2018-06-16)

- Add `partner` and `seller_id` options for legacy API, thanks @KQyongzhang #83

## v0.15.0 (2017-08-27)

- Add `Alipay::Client` for open API.
- Stop adding new feature to Alipay::Service, Alipay::App::Service, Alipay::Wap::Service, Alipay::Mobile::Service.

## v0.14.0 (2016-12-28)

- Add `Alipay::App::Service.create_forex_trade_wap_url` method, thanks @xiaohesong #61

## v0.13.0 (2016-12-24)

- Add `Alipay::App::Service.alipay_trade_app_pay` method, thanks @FX-HAO #64

## v0.12.0 (2016-03-07)

- Add `Alipay::Service::batch_trans_notify_url` method, thanks @ryancheung #58

## v0.11.0 (2015-12-04)

- Add `Alipay::Wap::Service.security_risk_detect` method, thanks @jasl #55

## v0.10.0 (2015-11-03)

- Add `Alipay::Service.account_page_query` method, thanks @xjz19901211 #53

## v0.9.0 (2015-10-15)

- Add `Alipay::Service.create_direct_pay_by_user_wap_url` method, thanks @serco-chen #52

## v0.8.0 (2015-07-20)

- Add `Alipay::Mobile::Service.mobile_security_pay_string` method, thanks @Sen #49
- Remove `Alipay::Service.mobile_security_pay_url` #49

## v0.7.1 (2015-04-10)

- Don't warning when `rmb_fee` is used in forex_refund_url method.

## v0.7.0 (2015-04-07)

- Remove Alipay.seller_email setting, it can be replaced by seller_id, than same with pid.
- Alipay::Wap::Service.trade_create_direct_token add a required params: seller_account_name.


## v0.6.0 (2015-04-06)

New API:

- Add Alipay::Service.mobile_security_pay_url.
- All service methods accept `options` args to config pid, seller_email and key.

New Config:

- Add `Alipay.sign_type`, default is 'MD5', allow: 'MD5', 'RSA'. 'DSA' is not implemented yet.

Break Changes:

- Move `Alipay::Service::Wap` to `Alipay::Wap::Service`
- Move `Alipay::Sign::Wap` to `Alipay::Wap::Sign`
- Move `Alipay::Notify::Wap` to `Alipay::Wap::Notify`
- Rename `Alipay::Service.create_forex_single_refund_url` to `Alipay::Service.forex_refund_url`
- Rename `Alipay::Service.create_forex_trade` to `Alipay::Service.create_forex_trade_url`
- Rename `Alipay::Service.create_refund_url` to `Alipay::Service.refund_fastpay_by_platform_pwd_url`
- Rename `Service::Wap.auth_and_execute` to `Service::Wap::Service.auth_and_execute_url`

Now `Alipay::Sign.verify?` and `Alipay::Wap::Sign.verify?` detect sign_type by params.

Development:

- Update Test::Unit to Minitest.
- Use fxied test data.

## v0.5.0 (2015-03-09)

- Add `forex_single_refund` service.
- Add `debug_mode` config:

  set `Alipay.debug_mode = false` to disable options check warning in production env.

## v0.4.1 (2015-03-03)

- Fix `single_trade_query` check options typo.

## v0.4.0 (2015-01-23)

- Add `single_trade_query` service. #19

## v0.3.1 (2015-01-15)

- Fix xml encoding #18

## v0.3.0 (2014-12-10)

- Add `close_trade` service. by @linjunpop #16

## v0.2.0 (2014-12-03)

- Add `create_forex_trade` service. by @christophe-dufour #15

## v0.1.0 (2014-06-15)

- Add Wap API by @HungYuHei

- Add App API by @HungYuHei


lib/alipay/client.rb
module Alipay
  class Client
    # Create a client to manage all API request.
    #
    # Example:
    #
    #   alipay_client = Alipay::Client.new(
    #     url: 'https://openapi.alipaydev.com/gateway.do',
    #     app_id: '2016000000000000',
    #     app_private_key: APP_PRIVATE_KEY,
    #     alipay_public_key: ALIPAY_PUBLIC_KEY
    #   )
    #
    # Options:
    #
    # [:url]  Alipay Open API gateway,
    #         'https://openapi.alipaydev.com/gateway.do'(Sandbox) or
    #         'https://openapi.alipay.com/gateway.do'(Production).
    #
    # [:app_id] Your APP ID.
    #
    # [:app_private_key] APP private key.
    #
    # [:alipay_public_key] Alipay public key.
    #
    # [:format] default is 'json', only support 'json'.
    #
    # [:charset] default is 'UTF-8', only support 'UTF-8'.
    #
    # [:sign_type] default is 'RSA2', support 'RSA2', 'RSA', 'RSA2' is recommended.
    def initialize(options)
      options = ::Alipay::Utils.stringify_keys(options)
      @url = options['url']
      @app_id = options['app_id']
      @app_private_key = options['app_private_key']
      @alipay_public_key = options['alipay_public_key']
      @format = options['format'] || 'json'
      @charset = options['charset'] || 'UTF-8'
      @sign_type = options['sign_type'] || 'RSA2'
      @app_cert_sn = options['app_cert_sn']
      @alipay_root_cert_sn = options['alipay_root_cert_sn']
    end

    # Generate a query string that use for APP SDK excute.
    #
    # Example:
    #
    #   alipay_client.sdk_execute(
    #     method: 'alipay.trade.app.pay',
    #     biz_content: {
    #       out_trade_no: '20160401000000',
    #       product_code: 'QUICK_MSECURITY_PAY',
    #       total_amount: '0.01',
    #       subject: 'test'
    #     }.to_json(ascii_only: true),
    #     timestamp: '2016-04-01 00:00:00'
    #   )
    #   # => 'app_id=2016000000000000&charset=utf-8&sig....'
    def sdk_execute(params)
      params = prepare_params(params)

      URI.encode_www_form(params)
    end

    # Generate a url that use to redirect user to Alipay payment page.
    #
    # Example:
    #
    #   alipay_client.page_execute_url(
    #     method: 'alipay.trade.page.pay',
    #     biz_content: {
    #       out_trade_no: '20160401000000',
    #       product_code: 'FAST_INSTANT_TRADE_PAY',
    #       total_amount: '0.01',
    #       subject: 'test'
    #     }.to_json(ascii_only: true),
    #     timestamp: '2016-04-01 00:00:00'
    #   )
    #   # => 'https://openapi.alipaydev.com/gateway.do?app_id=2016...'
    def page_execute_url(params)
      params = prepare_params(params)

      uri = URI(@url)
      uri.query = URI.encode_www_form(params)
      uri.to_s
    end

    # Generate a form string that use to render in view and auto POST to
    # Alipay server.
    #
    # Example:
    #
    #   alipay_client.page_execute_form(
    #     method: 'alipay.trade.page.pay',
    #     biz_content: {
    #       out_trade_no: '20160401000000',
    #       product_code: 'FAST_INSTANT_TRADE_PAY',
    #       total_amount: '0.01',
    #       subject: 'test'
    #     }.to_json(ascii_only: true),
    #     timestamp: '2016-04-01 00:00:00'
    #   )
    #   # => '<form id='alipaysubmit' name='alipaysubmit' action=...'
    def page_execute_form(params)
      params = prepare_params(params)

      html = %Q(<form id='alipaysubmit' name='alipaysubmit' action='#{@url}?charset=#{@charset}' method='POST'>)
      params.each do |key, value|
        html << %Q(<input type='hidden' name='#{key}' value='#{value.gsub("'", "&apos;")}'/>)
      end
      html << "<input type='submit' value='ok' style='display:none'></form>"
      html << "<script>document.forms['alipaysubmit'].submit();</script>"
      html
    end

    # Immediately make a API request to Alipay and return response body.
    #
    # Example:
    #
    #   alipay_client.execute(
    #     method: 'alipay.data.dataservice.bill.downloadurl.query',
    #     biz_content: {
    #       bill_type: 'trade',
    #       bill_date: '2016-04-01'
    #     }.to_json(ascii_only: true)
    #   )
    #   # => '{ "alipay_data_dataservice_bill_downloadurl_query_response":{...'
    def execute(params)
      params = prepare_params(params)

      Net::HTTP.post_form(URI(@url), params).body
    end

    # Generate sign for params.
    def sign(params)
      string = params_to_string(params)

      case @sign_type
      when 'RSA'
        ::Alipay::Sign::RSA.sign(@app_private_key, string)
      when 'RSA2'
        ::Alipay::Sign::RSA2.sign(@app_private_key, string)
      else
        raise "Unsupported sign_type: #{@sign_type}"
      end
    end

    # Verify Alipay notification.
    #
    # Example:
    #
    #   params = {
    #     out_trade_no: '20160401000000',
    #     trade_status: 'TRADE_SUCCESS'
    #     sign_type: 'RSA2',
    #     sign: '...'
    #   }
    #   alipay_client.verify?(params)
    #   # => true / false
    def verify?(params)
      params = Utils.stringify_keys(params)
      return false if params['sign_type'] != @sign_type

      sign = params.delete('sign')
      # sign_type does not use in notify sign
      params.delete('sign_type')
      string = params_to_string(params)
      case @sign_type
      when 'RSA'
        ::Alipay::Sign::RSA.verify?(@alipay_public_key, string, sign)
      when 'RSA2'
        ::Alipay::Sign::RSA2.verify?(@alipay_public_key, string, sign)
      else
        raise "Unsupported sign_type: #{@sign_type}"
      end
    end

    private

    def prepare_params(params)
      params = {
        'app_id' => @app_id,
        'charset' => @charset,
        'sign_type' => @sign_type,
        'version' => '1.0',
        'timestamp' => Time.now.localtime('+08:00').strftime("%Y-%m-%d %H:%M:%S")
      }.merge(::Alipay::Utils.stringify_keys(params))
      if !@app_cert_sn.nil? && !@alipay_root_cert_sn.nil?
        params = params.merge({
          'app_cert_sn' => @app_cert_sn,
          'alipay_root_cert_sn' => @alipay_root_cert_sn
        })
      end
      params['sign'] = sign(params)
      params
    end

    def params_to_string(params)
      params.sort.map { |item| item.join('=') }.join('&')
    end
  end
end
